addons/portal/controllers/thread.py

```diff
@@ -0,0 +1,17 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.http import request
+from odoo.addons.mail.controllers import thread
+
+
+class ThreadController(thread.ThreadController):
+
+    def _prepare_post_data(self, post_data, thread, **kwargs):
+        post_data = super()._prepare_post_data(post_data, thread, **kwargs)
+        token, _hash, pid = kwargs.get("token"), kwargs.get("hash"), kwargs.get("pid")
+        if (token or (_hash and pid)) and request.env.user._is_public() and thread.env.su:
+            if _hash and pid:
+                post_data["author_id"] = int(pid)
+            elif hasattr(thread, "partner_id") and thread.partner_id:
+                post_data["author_id"] = thread.partner_id.id
```

Comment:
It would be much safer and not harder to code to just check the hash/pid or token here again just to be safe.

Then the strange check on `env.su` can be removed, and then we don't rely on weak heuristics of "how we reached this code" to make it safe.

Just make two simple util functions to validate `thread, hash, pid` or `thread, token` and call it both here and in `_get_thread_with_access`.

I believe this will make both here and `_get_thread_with_access` easier to read, by hiding the comparison logic in a dedicated place.